lib/drivers/bluetoothLE/bluetoothLEDriver.js

```diff
@@ -0,0 +1,196 @@
+ * not, you can obtain one from Tidepool Project at tidepool.org.
+ * == BSD2 LICENSE ==
+ */
+
+import _ from 'lodash';
+import sundial from 'sundial';
+
+import TZOUtil from '../../TimezoneOffsetUtil';
+import annotate from '../../eventAnnotations';
+import { remote } from 'electron';
+
+const isBrowser = typeof window !== 'undefined';
+// eslint-disable-next-line no-console
+const debug = isBrowser ? require('bows')('BluetoothLEDriver') : console.log;
+
+module.exports = (config) => {
+  const cfg = _.clone(config);
+  cfg.deviceTags = ['bgm'];
+
+  // With no date & time settings changes available,
+  // timezone is applied across-the-board
+  cfg.tzoUtil = new TZOUtil(cfg.timezone, new Date().toISOString(), []);
+
+  return {
+   /* eslint no-param-reassign:
+      [ "error", { "props": true, "ignorePropertyModificationsFor": ["data"] } ] */
+
+   detect(deviceInfo, cb) {
+     debug('no detect function needed', deviceInfo);
+     cb(null, deviceInfo);
+   },
+
+   setup(deviceInfo, progress, cb) {
+     debug('in setup!');
+     progress(100);
+     cb(null, { deviceInfo });
+   },
+
+   async connect(progress, data, cb) {
+     try {
+       debug('in connect!');
+       await cfg.deviceComms.ble.connectTimeout();
+     } catch (error) {
+       return cb(error, null);
+     }
+     return cb(null, data);
+   },
+
+   async getConfigInfo(progress, data, cb) {
+     debug('in getConfigInfo', data);
+     progress(0);
+
+     try {
+       _.assign(cfg.deviceInfo , await cfg.deviceComms.ble.getDeviceInfo());
+       cfg.deviceTags = ['bgm'];
+       cfg.deviceInfo.deviceId = `${[cfg.deviceInfo.manufacturers]}-${cfg.deviceInfo.model}-${remote.getGlobal('bluetoothDeviceId')}`;
+       cfg.builder.setDefaults({ deviceId: cfg.deviceInfo.deviceId });
+
+       cfg.deviceComms.ble.once('numberOfRecords', (result) => {
+         debug('Number of records:', result);
+       });
+       await cfg.deviceComms.ble.getNumberOfRecords();
+     } catch (error) {
+       return cb(error, null);
+     }
+
+     return cb(null, data);
+   },
+
+   async fetchData(progress, data, cb) {
+     debug('in fetchData', data);
+
+     cfg.deviceComms.ble.once('data', (result) => {
+       debug('Records:', result);
+       data.records = result;
+       return cb(null, data);
+     });
+
+     try {
+       await cfg.deviceComms.ble.getAllRecords();
+     } catch (error) {
+       return cb (error, null);
+     }
+   },
+
+   processData(progress, data, cb) {
+     progress(0);
+     data.post_records = [];
+
+     _.forEach(data.records, (result) => {
+       const record = result;
+       let annotation = null;
+
+       if (record.value > 600) {
+         record.value = 601;
+         annotation = {
+           code: 'bg/out-of-range',
+           value: 'high',
+         };
+       } else if (record.value < 20) {
+         record.value = 19;
+         annotation = {
+           code: 'bg/out-of-range',
+           value: 'low',
+         };
+       }
```

Comment:
Are these HI and LO values consistent for all potentially supported BG meters? If not, should they be part of the device configuration to be more specific?

Reply:
This is something I wish was in the BLE glucose spec, so that meters would have to report their HI/LO thresholds, as it turns out none of them do.

Almost all the meters we currently support in the Uploader have a 20/600 HI/LO threshold, but if we do find one that differs, I'll have to figure out how to update this logic. I did check that the CareSens N Premier BLE meter, which is the first meter to use this driver, does indeed have 20/600 thresholds.

Ideally meters would report +/- infinity when it's out-of-range, right? Unfortunately they don't always do that, and you don't want to record a value that you know is out of spec for that meter (sometimes the only place you can find the actual thresholds is in the meter user manual).